feat: support HTTPS backends via certificate_transfer and tls-certificates relation (ISD-6137) (#141)
canonical/content-cache-operator: docs/how-to/enable-https.md

When the URL scheme is `https`, nginx connects to the backend over TLS on the specified port.

> ## Provide a CA certificate via certificate_transfer

Reviewer: ```suggestion
## Provide a CA certificate
```
For conciseness, since `certificate_transfer` isn't explicitly mentioned in the section

To verify the backend TLS certificate, integrate a certificate provider charm (such as
`self-signed-certificates` or `lego`) using the `receive-ca-cert` endpoint:

```bash
juju integrate <cert-provider>:send-ca-cert content-cache:receive-ca-cert
```

Once the CA certificate is received, the content-cache charm will:

- Write the certificate to `/etc/nginx/certs/ca-bundle.pem`
- Configure nginx to verify backend certificates against this CA
- Use `proxy_ssl_verify on` and `proxy_ssl_server_name off` (SNI hostname validation is

>   disabled in this version)

Author: Fixed — changed to 'disables SNI' so the wording accurately reflects what `proxy_ssl_server_name off` does.

> If HTTPS backends are configured but no CA certificate has been provided, the charm enters

Reviewer: ```suggestion
If HTTPS backends are configured but no CA certificate has been provided, the charm will enter
```

## TLS termination for incoming traffic

When HAProxy connects to the content-cache over HTTPS, the charm must present a TLS
certificate. This is configured through the `certificates` relation
(interface: `tls-certificates`).

Deploy a TLS certificate provider (e.g. `lego`) as `cache-lego` and integrate:

```bash
juju integrate content-cache:certificates cache-lego:certificates
```

When the certificate is issued, the charm automatically:

1. Writes the combined cert+key PEM to `/etc/nginx/certs/<unit-ip>.pem`
2. Reconfigures nginx to listen with `ssl` on the allocated port

> 3. Updates `cache-backends` to return `https://` URLs

Reviewer: A numbered list implies an order to the operations. Just want to check that these steps are always followed in this exact order -- otherwise we probably want to use a bulleted list instead

Author: The order is always sequential: nginx must be reconfigured (step 2) before the charm can write `cache-backends` (step 3), and the cert file must exist before nginx reconfiguration (step 1). Keeping the numbered list.

> 3. Updates `cache-backends` to return `https://` URLs

Reviewer: ```suggestion
3. Updates the `cache-backends` relation data to return `https://` URLs
```

HAProxy must trust this certificate. Integrate HAProxy with `cache-lego` using the
`certificate_transfer` interface:

```bash
juju integrate cache-lego:send-ca-cert haproxy:receive-ca-cert
```

If the `certificates` relation is present but the certificate has not yet been issued,
the charm enters `WaitingStatus`. If the relation is removed, the charm deletes the

> cert file and reverts nginx to HTTP automatically.

Reviewer: ```suggestion
the charm enters `WaitingStatus`. If the relation is removed, the charm automatically deletes the
cert file and reverts nginx to HTTP.
```
I assumed here that the file deletion is also automatic
